File: Trumpet.Backend/Services/CommunityItemsService.cs
using Microsoft.EntityFrameworkCore;
using Trumpet.Backend.Data;
using Trumpet.Backend.Models;

namespace Trumpet.Backend.Services.CommunityItems;

public class CommunityItemsService : ICommunityItemsService
{
    private readonly MusicContext _context;

    public CommunityItemsService(MusicContext context)
    {
        _context = context;
    }

    public async Task<IEnumerable<Item>> GetItemsByCommunityIdAsync(string communityId)
    {
        Console.WriteLine($"[CommunityItemsService] Getting items for communityId: {communityId}");
        
        // Get all sub-community IDs recursively
        var allCommunityIds = new List<string> { communityId };
        var currentLevelIds = new List<string> { communityId };

        while (currentLevelIds.Any())
        {
            var searchIds = currentLevelIds;
            var nextLevelIds = await _context.Communities
                .Where(c => c.ParentCommunityId != null && searchIds.Contains(c.ParentCommunityId))
                .Select(c => c.Id)
                .ToListAsync();
            
            if (!nextLevelIds.Any()) break;

            allCommunityIds.AddRange(nextLevelIds);
            currentLevelIds = nextLevelIds;
        }

        Console.WriteLine($"[CommunityItemsService] All community IDs (including sub-communities): {string.Join(", ", allCommunityIds)}");

        // Check what collections exist for these communities
        var collections = await _context.Collections
            .Where(c => c.ParentCommunityId != null && allCommunityIds.Contains(c.ParentCommunityId))
            .ToListAsync();
        
        Console.WriteLine($"[CommunityItemsService] Found {collections.Count} collections for these communities:");
        foreach (var col in collections)
        {
            Console.WriteLine($"  - Collection: {col.Id} | Name: {col.Name} | ParentCommunityId: {col.ParentCommunityId}");
        }

        // Check all collections to see what ParentCommunityIds they have
        var allCollections = await _context.Collections.Take(10).ToListAsync();
        Console.WriteLine($"[CommunityItemsService] Sample of all collections (first 10):");
        foreach (var col in allCollections)
        {
            Console.WriteLine($"  - Collection: {col.Id} | ParentCommunityId: {col.ParentCommunityId ?? "NULL"}");
        }

        // Fetch items from collections belonging to any of the found communities
        var items = await _context.Items
            .Include(i => i.Metadata)
            .Include(i => i.Bitstreams)
            .Where(i => i.CollectionId != null && _context.Collections.Any(c => allCommunityIds.Contains(c.ParentCommunityId) && c.Id == i.CollectionId))
            .OrderBy(i => i.Name)
            .ToListAsync();

        Console.WriteLine($"[CommunityItemsService] Returning {items.Count} items");
        
        return items;
    }
}

File: Backend/src/Services/CommunityItemsService.php
<?php

declare(strict_types=1);

namespace Trumpet\Services;

use PDO;

/**
 * CommunityItemsService
 *
 * Mirrors community_items_service.py — recursively collects all
 * sub-community IDs, then fetches items from their collections.
 */
class CommunityItemsService
{
    public function __construct(private readonly PDO $pdo)
    {
    }

    public function getItemsByCommunityId(string $communityId): array
    {
        // ── 1. Recursively collect all descendant community IDs ──────────────
        $allCommunityIds   = [$communityId];
        $currentLevelIds   = [$communityId];

        while (!empty($currentLevelIds)) {
            $placeholders = implode(',', array_fill(0, count($currentLevelIds), '?'));
            $stmt = $this->pdo->prepare(
                "SELECT Id FROM Communities WHERE ParentCommunityId IN ({$placeholders})"
            );
            $stmt->execute($currentLevelIds);
            $nextLevelIds = $stmt->fetchAll(PDO::FETCH_COLUMN);

            $newIds = array_diff($nextLevelIds, $allCommunityIds);
            if (empty($newIds)) {
                break;
            }

            $allCommunityIds = array_merge($allCommunityIds, array_values($newIds));
            $currentLevelIds = array_values($newIds);
        }

        // ── 2. Fetch items whose collection belongs to any of those communities
        $placeholders = implode(',', array_fill(0, count($allCommunityIds), '?'));
        $sql = "
            SELECT i.Id, i.Name, i.Handle, i.LastModified, i.Withdrawn, i.Archived, i.CollectionId
            FROM Items i
            JOIN Collections c ON i.CollectionId = c.Id
            WHERE c.ParentCommunityId IN ({$placeholders})
            ORDER BY i.Name
        ";

        $stmt = $this->pdo->prepare($sql);
        $stmt->execute($allCommunityIds);
        $items = $stmt->fetchAll();

        // ── 3. Hydrate metadata + bitstreams (reuse ItemsService logic) ──────
        $itemsService = new ItemsService($this->pdo);
        return $itemsService->hydratePublic($items);
    }
}
